Update/overhaul documentation

willkg opened the pull request:
This updates and overhauls the documentation:

1. reorganizes the documentation into "user", "dev/deploy", and "theory" sections
2. cleans up API documentation and makes deprecations clearer
3. creates a separate data flows document
4. updates the Sphinx makefile to something a current `sphinx-quickstart` would create
5. removes the old dev documentation

There's some new FIXME marks for things that should probably be written. I think there's still some stuff that could get updated, but we can do that in another pass.

To test, do `make docs` and then read through things

jwhitlock commented on docs/algo/accuracy.rst:
```diff
@@ -6,102 +6,101 @@ Accuracy
-only partially expose the cell identifiers, for example only exposing
-the carrier and primary scrambling code of the neighboring cells.
+limitations in chipset drivers, the radio interface layer and the operating
+systems often hide this data from application code. Or only partially expose
```
"Or only partially expose" is a sentence fragment, not a complete sentence.

willkg commented:
I finished doing a language pass. While doing that, I removed a lot of flourish-type words to improve clarity with less text, added some details, fixed some formatting, and changed some markup.

I also kind of doubled the size of this PR. Oops. :sob: 

This is a big PR, so I'm going to cap it to one or two rounds of review-and-fix and land whatever we end up with before end of Friday. After that, we can do additional fixes in future PRs.

@jwhitlock It's ready for reviewing.

jwhitlock commented on docs/dataflows.rst:
```diff
@@ -0,0 +1,134 @@
+
+* ``blue_shard_*``
+* ``cell_*``
+* ``wifi_shard_*``
```
It may be useful to specify the range of those wildcards:

* blue_shard_*: hex 0 through f
* cell_*: gsm, lte, wcdma, area?
* wifi_shared_*: hex 0 through f

willkg replied:
Good idea!

jwhitlock commented on docs/dataflows.rst:
```diff
@@ -0,0 +1,134 @@
+   If the query is handled by ``InternalPositionSource``, then the web frontend
+   adds a submission.
+
+2. If the sumission is kept, then the web frontend adds an item to the
```
```suggestion
2. If the submission is kept, then the web frontend adds an item to the
```

willkg replied:
:+1:

jwhitlock commented on docs/dataflows.rst:
```diff
@@ -0,0 +1,134 @@
+   "source" can be one of:
+   
+   * ``gnss``: Global Navigation Satellite System based data
+   * ``fused``: observation data positioned based on fused data
```
[geosubmit2](https://ichnaea.readthedocs.io/en/latest/api/geosubmit2.html) suggests:
```suggestion
   * ``fused``: a position obtained from a combination of other sensors or outside service queries
```

willkg replied:
:+1:

jwhitlock commented on docs/dataflows.rst:
```diff
@@ -0,0 +1,134 @@
+   * ``query``: position estimate based on query data
+
+3. The Celery scheduler schedules the ``update_incoming`` task every 
+   X seconds--see task definition in `ichnaea/data/tasks.py
```
```suggestion
   32 seconds--see task definition in `ichnaea/data/tasks.py
```

willkg replied:
I didn't want to duplicate the number in two places because it's such an odd number I figured we'd change it. That's why I pointed to the code. I think I want to leave it as X for now.

jwhitlock commented on docs/dataflows.rst:
```diff
@@ -0,0 +1,134 @@
+     
+     This data ends up in Redis queues like `update_cell_gsm`,
+     `update_cell_wcdma`, `update_wifi_0`, `update_wifi_1`, etc to be
+     processed by `station updater` tasks.
```
Some more fixed width styling like ``update_wifi_0`` here and in the next paragraph would make this more consistent with other content.

willkg replied:
:+1:

jwhitlock commented on docs/dataflows.rst:
```diff
@@ -0,0 +1,134 @@
+Datamap data
+============
+
+Datamap data is stored in the ``datamap_*`` tables.
```
If you list the ranges on ``blue_shard_*``, then this may need it too. The values are ``ne``, ``nw``, ``se``, and ``sw``.

willkg replied:
:+1:

jwhitlock commented on docs/debug.rst:
```diff
@@ -4,65 +4,20 @@
 
-First, lets check the database configuration. If you have a development
-setup, you can open a MySQL client prompt via:
+First, lets check the database configuration.
```
```suggestion
First, let's check the database configuration.
```

willkg replied:
:+1:

jwhitlock commented on docs/debug.rst:
```diff
@@ -4,65 +4,20 @@
 .. code-block:: bash
 
-    docker-compose exec mysql mysql -uroot -plocation location
+   make mysql
```
Nice, didn't know this one 😀

willkg replied:
It might be new-ish. I forget when I added it. I'm wondering if we should add one for redis-cli, too.

jwhitlock commented on docs/debug.rst:
```diff
@@ -74,9 +29,8 @@ First up, check if the alembic database setup and migration was run:
-This needs to produce a single row and some `version_num` in it.
-If it isn't there, check the `Database Setup` part of
-:ref:`the deploy docs <deploy>`.
+This needs to produce a single row and some `version_num` in it.  If it isn't
```
```suggestion
This needs to produce a single row and some ``version_num`` in it.  If it isn't
```

willkg replied:
:+1:

jwhitlock commented on docs/debug.rst:
```diff
@@ -111,84 +65,74 @@ And the export config:
 .. code-block:: bash
 
-    redis-cli -h $REDIS_HOST
+    $ redis-cli -u $REDIS_URI
```
Nice!

willkg replied:
I'm pretty happy with this, but tossing around creating a make rule for it.

jwhitlock commented on docs/debug.rst:
```diff
@@ -203,26 +147,29 @@ and inside it run:
 
-Now that all the building blocks are in place, let's try to send them
-real data to the service and see how it processes it.
+Now that all the building blocks are in place, let's try to send them real data
```
```suggestion
Now that all the building blocks are in place, let's try to send real data
```

willkg replied:
:+1:

jwhitlock commented on docs/index.rst:
```diff
@@ -19,35 +18,36 @@ You can interact with the service in two ways:
+About the name
+==============
+
+In Greek mythology, Ichnaea (Iknaia) means "the tracker".
```
I'm glad this detail survived, and this is a better place for it. We could throw in the Greek (Ιχναίη from https://en.wikipedia.org/wiki/Ichnaea) and get someone to pronounce it for us at the next All Hands.

jwhitlock commented on docs/devdeploy_index.rst:
```diff
@@ -0,0 +1,18 @@
+====================================
+Development/Deployment documentation
```
This is an improvement over all of this coming under installation 👍

willkg commented on docs/api/index.rst:
```diff
@@ -38,8 +37,10 @@ it might take longer.
-distributing API keys.
+.. Note::
+
+   Mozilla is currently evaluating its MLS service and terms and is not
```
I like the idea, but I'm going to punt on this for now.

willkg commented on docs/algo/accuracy.rst:
```diff
@@ -4,104 +4,102 @@
-tablets. So movement detection and detecting inconsistencies between
-observed data and the database world view are important considerations.
+Bluetooth and WiFi networks have a fairly limited range. Bluetooth low-energy
+beacons typically just a couple meters and WiFi networks up to 100 meters. With
```
:+1:

willkg commented on docs/algo/accuracy.rst:
```diff
@@ -4,104 +4,102 @@
+
+One difficult challenge with Bluetooth and WiFi networks are the constantly
+moving networks. For example, WiFi networks installed on buses or trains or in
+the form of hotspot-enabled mobile phones or tablets. Detection movement and
```
:+1:

willkg commented on docs/algo/accuracy.rst:
```diff
@@ -4,104 +4,102 @@
+In GSM networks, one typically only has access to the unique cell id of the
+serving cell. In GSM networks, the phone does not know the full cell ids of any
+neighboring cells unless it associates with the new cell as part of a hand-over
+and forgetting the cell id of the old cell.
```
:+1:

willkg commented on docs/algo/accuracy.rst:
```diff
@@ -4,104 +4,102 @@
-In more populated places the cell sizes are typically much smaller,
-but generally accuracy will be in the tens of kilometer range.
+In more populated places the cell sizes are typically much smaller, but
+accuracy will be in the tens of kilometer range.
```
:+1:

willkg commented on docs/algo/accuracy.rst:
```diff
@@ -4,104 +4,102 @@
+In WCDMA networks, neighboring cell information can be available. However,
+limitations in chipset drivers, the radio interface layer, and the operating
+systems often hide this data from application code or only partially expose the
+cell identifiers. For example, they might only exposing the carrier and primary
```
:+1:

willkg commented on docs/algo/accuracy.rst:
```diff
@@ -4,104 +4,102 @@
 
-In urban areas we should typically see accuracy in the 1 to 10 kilometer
-range.
+In urban areas we should typically see accuracy in the 1 to 10 kilometer range.
```
:+1:

willkg commented on docs/algo/accuracy.rst:
```diff
@@ -4,104 +4,102 @@
-the world GeoIP is only accurate to the region level.
+The accuracy of GeoIP depends on the region the user is in. In the US, GeoIP
+can be fairly accurate and often places the user in the right city or
+metropolitan area. In many other parts of the world GeoIP is only accurate to
```
:+1:

willkg commented on docs/algo/accuracy.rst:
```diff
@@ -4,104 +4,102 @@
-the timing advance. While we currently don't use this information, in
-the future it has the potential to significantly improve position
-estimates based on multiple cells.
+LTE networks also expose a time based distance metric in the form of the timing
```
:+1:

willkg commented on docs/algo/index.rst:
```diff
@@ -9,38 +9,38 @@ The project uses a couple of different approaches and algorithms.
-2. Generate signal fingerprints for a fine-grained grid of the world.
-   Find best match for a observed fingerprint.
+2. Generate signal fingerprints for a fine-grained grid of the world. Find
+   best match for an observed fingerprint.
```
:+1:

willkg commented on docs/algo/index.rst:
```diff
@@ -9,38 +9,38 @@ The project uses a couple of different approaches and algorithms.
-infer a distance measure from it. The further a device is away from the
-signal source, the weaker the signal should get.
+In theory one could use signal strength data to infer a distance measure from
+it: the further a device is away from the signal source, the weaker the signal
```
:+1:

willkg commented on docs/algo/index.rst:
```diff
@@ -9,38 +9,38 @@ The project uses a couple of different approaches and algorithms.
+accurate distance predictor. Fine grained enough timing data is unfortunately
+almost never available to the application or operating system layer in client
+devices. Some LTE networks and really modern WiFi networks with support for
+802.11v are the rare exception to this. These are so rare, that we currently
```
:+1:

willkg commented on docs/api/geolocate.rst:
```diff
@@ -1,30 +1,32 @@
-Geolocate
-=========
+========================
+Geolocate: /v1/geolocate
```
Me, too. I was having troubles matching API endpoints to monikers.

willkg commented on docs/api/geolocate.rst:
```diff
@@ -156,11 +156,11 @@ signalStrength
 radioType
-    The type of radio network. One of `gsm`, `wcdma` or `lte`.
-    This is a custom extension to the GLS API, which only defines the
+    The type of radio network. One of ``gsm``, ``wcdma``, or ``lte``.
```
Yeah... I almost didn't do this because it felt like I was imposing my style on things. Now that I'm looking at it, those should all be glossary term links.

willkg replied:
Future someone should do a pass on linking things to the glossary.

willkg commented on docs/api/geosubmit.rst:
```diff
@@ -1,23 +1,27 @@
     https://location.services.mozilla.com/v1/geosubmit?key=<API_KEY>
 
-Geosubmit requests are submitted using a POST request with a JSON body:
+Geosubmit requests are submitted using a POST request with a JSON body.
```
:+1:

willkg commented on docs/api/geosubmit2.rst:
```diff
@@ -1,24 +1,29 @@
+There is an earlier :ref:`api_geosubmit` v1 API, with a slightly different and
+less extensive field list.
+
+Geosubmit requests are submitted using a POST request with a JSON body.
```
:+1:

willkg commented on docs/api/geosubmit2.rst:
```diff
@@ -198,14 +203,16 @@ timingAdvance
-    The BSSID of the Wifi network. Hidden Wifi networks must not be collected.
+    The BSSID of the Wifi network.
+    
+    Hidden Wifi networks must not be collected.
```
Exciting stuff!

willkg commented on docs/api/submit.rst:
```diff
@@ -1,23 +1,28 @@
+=======
 
-Submit requests are submitted using a POST request to the following URL::
+Submit requests are submitted using a POST request to one of::
```
:+1:

willkg commented on docs/architecture.rst:
```diff
@@ -0,0 +1,79 @@
+========
+
+The application consists of an HTTP web service implementing the APIs and
+web site and a streaming data pipeline.
```
:+1:

willkg commented on docs/architecture.rst:
```diff
@@ -0,0 +1,79 @@
+.. _architecture:
+
+============
+Architecture
```
I think this was in there before, but I moved it and updated it.

willkg commented on docs/configure.rst:
```diff
@@ -145,110 +142,114 @@ to be set, for example:
-* Amazon S3 buckets
-* The projects own internal data processing pipeline
-* A HTTPS POST endpoint accepting the geosubmit v2 format
+* s3: Amazon S3 buckets
```
:+1:

willkg commented on docs/configure.rst:
```diff
@@ -145,110 +142,114 @@ to be set, for example:
-* The projects own internal data processing pipeline
-* A HTTPS POST endpoint accepting the geosubmit v2 format
+* s3: Amazon S3 buckets
+* internal: Ichnaea's internal data processing pipeline which creates/
```
:+1:

willkg commented on docs/configure.rst:
```diff
@@ -145,110 +142,114 @@ to be set, for example:
+* s3: Amazon S3 buckets
+* internal: Ichnaea's internal data processing pipeline which creates/
+  updates position data using new position information
+* geosubmit: submitting position information to an HTTP POST endpoint in
```
:+1:

willkg commented on docs/configure.rst:
```diff
@@ -145,110 +142,114 @@ to be set, for example:
-the ``skip_keys`` setting can be used to prevent data being
-round tripped and send back to the same partner that it came from.
+If the project is taking in data from a partner in a data exchange, the
+``skip_keys`` setting can be used to prevent data being round tripped and send
```
:+1:

willkg commented:
Thank you!
